core/drivers/hi16xx_uart.c

```diff
@@ -88,47 +86,29 @@ void hi16xx_uart_init(vaddr_t base, uint32_t uart_clk, uint32_t baud_rate)
 {
 	uint16_t freq_div = uart_clk / (16 * baud_rate);
 
-	/* Clear and enable FIFO */
-	write32(UART_FCR_FIFO_EN | UART_FCR_RX_FIFO_RST | UART_FCR_TX_FIFO_RST,
-		base + UART_FCR);
-	dsb();
+	/* Enable (and clear) FIFOs */
+	write32(UART_FCR_FIFO_EN, base + UART_FCR);
 
 	/* Enable access to _DLL and _DLH */
 	write32(UART_LCR_DLAB, base + UART_LCR);
-	dsb();
 
 	/* Calculate and set UART_DLL */
 	write32(freq_div & 0xFF, base + UART_DLL);
-	dsb();
 
 	/* Calculate and set UART_DLH */
 	write32((freq_div >> 8) & 0xFF, base + UART_DLH);
-	dsb();
 
-	/*
-	 * Clear _DLL and _DLH access bit, set data size (8 bits), parity etc.
-	 */
+	/* Clear _DLL/_DLH access bit, set data size (8 bits), parity etc. */
 	write32(UART_LCR_DLS8, base + UART_LCR);
-	dsb();
 
 	/* Disable interrupt mode */
 	write32(0, base + UART_IEL);
-	dsb();
```

Comment:
hmmm, you may be right.
i will try to dig...

... actually, after few explorations, yes you are right.
StronglyOrdered and Device memories are both non-reorderable (sic!) on
armv7.

However, I found something interesting.
> Device-nGnRnE most restrictive (equivalent to Strongly Ordered memory in
the ARMv7 architecture).
[armv8-A prog. guide|
https://developer.arm.com/docs/den0024/latest/13-memory-ordering/131-memory-types/1312-device-memory
]

Hence, maybe the "device" mapping applied in optee (DevicenGnRnE) is a bit
rude and should be set to Device-nGnRE ?


On 11 August 2016 at 17:05, David Brown <notifications@github.com> wrote:

> In core/drivers/hi16xx_uart.c
> <https://github.com/OP-TEE/optee_os/pull/977#discussion_r74439622>:
>
> >
> >  	/* Disable interrupt mode */
> >  	write32(0, base + UART_IEL);
> > -	dsb();
>
> I thought device memory couldn't be reordered, it just might be buffered.
> Strongly ordered would be similar to having a dsb() after every write, and
> makes it quite slow.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OP-TEE/optee_os/pull/977/files/83b87f2780461ecb754402aa50c8d4c08698b404#r74439622>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATEHJsNmWOLjK-I-uxP11Pe0OrxfGSNDks5qezo6gaJpZM4Jf2Dt>
> .
>